File: java/20_ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        if(s == null || s.length() == 0)
        	return true;
        Stack<Character> stack = new Stack<Character>();
        
        for(Character ch:s.toCharArray()) {
        	if(ch == '(') {
        		stack.push(')');
        	}else if(ch == '['){
				stack.push(']');
			}else if(ch == '{'){
				stack.push('}');
			}else {
				if(stack.isEmpty() || stack.pop() != ch) {
					return false;
				}
			}
        }
        
        return stack.isEmpty();
    }
}

File: c++/20_ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for (auto c : s) {
            if (c == '(' || c == '[' || c == '{') stk.push(c);
            else {
                if (c == ')' && stk.size() && stk.top() == '(') stk.pop();
                else if (c == ']' && stk.size() && stk.top() == '[') stk.pop();
                else if (c == '}' && stk.size() && stk.top() == '{') stk.pop();
                else return false;
            }
        }
        return stk.empty();
    }
};
